File: 15-3sum/15-3sum.java
// Last updated: 25/03/2026, 15:17:51

class Solution {
    public void twoSumHelper(int f, int[] nums, List<List<Integer>> res) {
        int i = f + 1,
            j = nums.length - 1;
        // [-1, -1, 0, 0, 1, 1]
        //.  f
        //.            i
        //                j
        // sum = 0
        // [[-1, 0, 1], [-1, 0, 1]]
        while (i < j) {
            int sum = nums[f] + nums[i] + nums[j];
            if (sum > 0) {
                j = j - 1;
            } else if (sum < 0) {
                i = i + 1;
            } else {
                // found a triplet
                // [-1, -1, 0, 0, 0, 0, 1, 1, 1, 1]
                //.                     f
                //.                        i
                //                               j
                // sum = 1
                // [[-1, 0, 1], [0, 0, 0]]
                res.add(Arrays.asList(nums[f], nums[i], nums[j])); // nums[f], nums[i], nums[j]
                i = i + 1;
                j = j - 1;
                while (i < j && nums[i] == nums[i - 1]) {
                    i = i + 1;
                }
                while (i < j && nums[j] == nums[j + 1]) {
                    j = j - 1;
                }
            }
        }
    }
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        // T: O(n.logn + n^2)
        // S: O(logn)
        // [-1, -1, 0, 0, 1, 1]
        //.  f
        // []
        List<List<Integer>> res = new ArrayList<>();
        for (int f = 0; f < nums.length; f = f + 1) {
            if (nums[f] > 0) {
                break;
            }
            if (f == 0 || nums[f] != nums[f - 1]) {
                 twoSumHelper(f, nums, res);
            }
        }
        return res;
    }
}

File: 15-3Sum/15-3Sum.cpp
// Last updated: 22/05/2026, 22:10:52
1class Solution {
2public:
3    vector<vector<int>> threeSum(vector<int>& nums) {
4        vector<vector<int>> ans;
5        sort(nums.begin(), nums.end());
6
7        int n = nums.size();
8
9        for(int i = 0; i < n; i++) {
10
11            // skip duplicate i
12            if(i > 0 && nums[i] == nums[i - 1]) continue;
13
14            int left = i + 1;
15            int right = n - 1;
16
17            while(left < right) {
18                int sum = nums[i] + nums[left] + nums[right];
19
20                if(sum == 0) {
21                    ans.push_back({nums[i], nums[left], nums[right]});
22
23                    left++;
24                    right--;
25
26                    // skip duplicate left
27                    while(left < right && nums[left] == nums[left - 1]) left++;
28
29                    // skip duplicate right
30                    while(left < right && nums[right] == nums[right + 1]) right--;
31                }
32                else if(sum < 0) {
33                    left++;
34                }
35                else {
36                    right--;
37                }
38            }
39        }
40
41        return ans;
42    }
43};